File: frontend/src/storage/readingProgress/document.ts
// The on-device reading-progress document used by browser and Wails desktop
// clients (the standalone reader included). Android keeps its existing per-book
// progress.json files because they are already app-private and scoped per
// connection.
//
// Every entry carries the wall-clock time it was written so that the desktop app
// and the standalone reader — two processes sharing one reading_progress.json —
// can reconcile concurrent writes newest-wins per book instead of by namespace.
// A reset is a timestamped tombstone (offset 0), not a deletion, so it too wins
// by recency rather than being resurrected by an older advance. Callers still
// deal only in plain offsets; the timestamp is internal to this module.

// Bumped from 1 to 2 for the timestamped shape. A document of any other version
// is treated as absent (see parse): the app is pre-alpha, so untimestamped v1
// progress is discarded rather than migrated.
export const READING_PROGRESS_DOCUMENT_VERSION = 2;

/** A book's stored progress: a UTF-16 offset and the epoch ms it was written. */
export interface ProgressEntry {
  offset: number;
  at: number;
}

export interface ReadingProgressDocument {
  version: number;
  /** device-document key -> book id -> progress entry. */
  shelves: Record<string, Record<string, ProgressEntry>>;
}

export function createReadingProgressDocument(): ReadingProgressDocument {
  return { version: READING_PROGRESS_DOCUMENT_VERSION, shelves: {} };
}

/** A non-negative, safe-integer number, or null when the value is not one. */
function normalizeCount(value: unknown): number | null {
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0) {
    return null;
  }
  return value;
}

// An entry is kept when its offset is a valid non-negative integer and it is
// meaningful: a positive offset, or a zero offset with a timestamp (a reset
// tombstone). A bare {offset:0, at:0} is nothing and is dropped. An unparseable
// timestamp is treated as unknown (0, the oldest) rather than dropping the entry.
function normalizeEntry(value: unknown): ProgressEntry | null {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    return null;
  }
  const record = value as Record<string, unknown>;
  const offset = normalizeCount(record.offset);
  if (offset === null) {
    return null;
  }
  const at = normalizeCount(record.at) ?? 0;
  if (offset === 0 && at === 0) {
    return null;
  }
  return { offset, at };
}

function normalizeShelf(value: unknown): Record<string, ProgressEntry> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    return {};
  }

  const books: Record<string, ProgressEntry> = {};
  for (const [bookID, entryValue] of Object.entries(value as Record<string, unknown>)) {
    const trimmedID = bookID.trim();
    const entry = normalizeEntry(entryValue);
    if (trimmedID && entry !== null) {
      books[trimmedID] = entry;
    }
  }
  return books;
}

/**
 * Reading progress is convenience state. A missing, corrupt, wrongly shaped,
 * or other-version document therefore falls back to an empty document rather
 * than preventing the reader from opening.
 */
export function parseReadingProgressDocument(
  text: string | null | undefined
): ReadingProgressDocument {
  if (typeof text !== 'string' || text.trim() === '') {
    return createReadingProgressDocument();
  }

  let raw: unknown;
  try {
    raw = JSON.parse(text);
  } catch {
    return createReadingProgressDocument();
  }

  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
    return createReadingProgressDocument();
  }

  const candidate = raw as Partial<ReadingProgressDocument>;
  if (candidate.version !== READING_PROGRESS_DOCUMENT_VERSION) {
    return createReadingProgressDocument();
  }

  const shelves: Record<string, Record<string, ProgressEntry>> = {};
  if (
    typeof candidate.shelves === 'object' &&
    candidate.shelves !== null &&
    !Array.isArray(candidate.shelves)
  ) {
    for (const [shelfKey, value] of Object.entries(candidate.shelves)) {
      const trimmedKey = shelfKey.trim();
      const books = normalizeShelf(value);
      if (trimmedKey && Object.keys(books).length > 0) {
        shelves[trimmedKey] = books;
      }
    }
  }

  return { version: READING_PROGRESS_DOCUMENT_VERSION, shelves };
}

export function serializeReadingProgressDocument(doc: ReadingProgressDocument): string {
  return JSON.stringify(doc);
}

export function getBookReadingOffset(
  doc: ReadingProgressDocument,
  shelfKey: string,
  bookID: string
): number {
  return doc.shelves[shelfKey]?.[bookID]?.offset ?? 0;
}

/**
 * The full stored entry (offset plus the epoch ms it was written) for a book, or
 * null when the book has no entry. Unlike getBookReadingOffset this also exposes
 * the timestamp, which a "last read" label needs; writers still go through the
 * offset helpers so the timestamp stays internal to writes.
 */
export function getBookReadingEntry(
  doc: ReadingProgressDocument,
  shelfKey: string,
  bookID: string
): ProgressEntry | null {
  return doc.shelves[shelfKey]?.[bookID] ?? null;
}

function setBookEntry(
  doc: ReadingProgressDocument,
  shelfKey: string,
  bookID: string,
  entry: ProgressEntry
): ReadingProgressDocument {
  return {
    ...doc,
    shelves: {
      ...doc.shelves,
      [shelfKey]: { ...(doc.shelves[shelfKey] ?? {}), [bookID]: entry }
    }
  };
}

/**
 * Records an offset for a book, stamped with `at` (defaulting to now) so a later
 * cross-process merge can pick the most recent write.
 *
 * A non-positive or invalid offset is a reset: it stores a zero-offset tombstone
 * carrying `at`, rather than deleting the entry, so the reset out-competes an
 * older advance the other process may hold. Resetting a book that has no entry at
 * all is a no-op — there is nothing to out-compete, and a tombstone would only be
 * noise. An unchanged offset is also a no-op, so re-reading the same position does
 * not churn the timestamp or force a write.
 */
export function withBookReadingOffset(
  doc: ReadingProgressDocument,
  shelfKey: string,
  bookID: string,
  offset: number,
  at: number = Date.now()
): ReadingProgressDocument {
  const trimmedShelfKey = shelfKey.trim();
  const trimmedBookID = bookID.trim();
  if (!trimmedShelfKey || !trimmedBookID) {
    return doc;
  }

  const previous = doc.shelves[trimmedShelfKey]?.[trimmedBookID];
  const normalizedOffset = normalizeCount(offset);

  // Reset: any non-positive or invalid offset. Missing entries already read as
  // zero, so only an existing, non-tombstone entry needs a tombstone written.
  if (normalizedOffset === null || normalizedOffset === 0) {
    if (!previous || previous.offset === 0) {
      return doc;
    }
    return setBookEntry(doc, trimmedShelfKey, trimmedBookID, { offset: 0, at });
  }

  if (previous && previous.offset === normalizedOffset) {
    return doc;
  }

  return setBookEntry(doc, trimmedShelfKey, trimmedBookID, { offset: normalizedOffset, at });
}

File: internal/readingprogress/document.go
// Package readingprogress models the cross-process reading-progress document
// and the reader -> desktop projection that folds standalone-reader progress
// back into the desktop library.
//
// Two independent processes share one reading_progress.json: the desktop app
// (which keys progress by a book's real shelf id) and the standalone reader. A
// reader launched from the desktop is handed the book's real shelf id and keys
// progress directly under it; a reader run on its own has no real shelf and keys
// everything under the synthetic shelf id "book", which the projection folds onto
// real shelves.
//
// Every entry carries the wall-clock time it was written, and all reconciliation
// is newest-wins per book: a concurrent write from either process is arbitrated
// by recency rather than by namespace ownership, so the reader and desktop can
// both write the same real shelf without losing each other's updates. A reset is
// a timestamped tombstone (offset 0) rather than a deletion, so it too wins by
// recency instead of being silently resurrected by an older advance. This package
// owns the document shape both sides agree on, the newest-wins merge, and the
// newest-wins projection. The file-backed, cross-process-locked store lives in
// store.go.
//
// The document shape mirrors frontend/src/storage/readingProgress/document.ts,
// which owns the single client-side implementation. This package only reads and
// writes what that format defines.
package readingprogress

import (
	"encoding/json/jsontext"
	"encoding/json/v2"
	"math"
	"strings"

	"github.com/voilelab/plainshelf/internal/jsonopt"
	"github.com/voilelab/plainshelf/internal/util"
)

// DocumentVersion matches READING_PROGRESS_DOCUMENT_VERSION in the frontend. A
// document of any other version is treated as absent rather than upgraded: the
// timestamped v2 format deliberately does not read the untimestamped v1 one.
// This file is device-local state, outside the on-disk format commitments, and
// dropping v1 shipped in v0.10.0 as a documented breaking change (CHANGELOG.md,
// docs/installation.md) — not a pattern to follow for shelf data.
const DocumentVersion = 2

// ReaderShelfID is the synthetic shelf key the standalone reader writes progress
// under. It is the canonical value; reader/readerapi.ShelfID must equal it (a
// guard test in the reader package enforces this).
const ReaderShelfID = "book"

// maxSafeInteger mirrors JavaScript's Number.MAX_SAFE_INTEGER: the frontend
// stores offsets and timestamps as JSON numbers, so anything larger could not
// round-trip.
const maxSafeInteger = 1<<53 - 1

// Entry is one book's stored progress: a UTF-16 character offset and the epoch
// milliseconds it was written. At arbitrates concurrent writes (newest wins); a
// zero Offset with a non-zero At is a reset tombstone.
type Entry struct {
	Offset int64 `json:"offset"`
	At     int64 `json:"at"`
}

// Document is the on-device reading-progress document: a shelf key maps to a set
// of book ids, each carrying an Entry. It never reaches the server and is not
// authoritative shelf data.
type Document struct {
	Version int                         `json:"version"`
	Shelves map[string]map[string]Entry `json:"shelves"`
}

// New returns an empty, current-version document.
func New() Document {
	return Document{Version: DocumentVersion, Shelves: map[string]map[string]Entry{}}
}

// Parse reads a stored document. A missing, corrupt, wrongly shaped, or
// non-current-version document yields an empty document rather than an error, so
// a bad file can never keep a reader from opening — this mirrors the frontend
// parser's leniency.
func Parse(text string) Document {
	doc := New()

	if strings.TrimSpace(text) == "" {
		return doc
	}

	var raw struct {
		Version int `json:"version"`
		Shelves map[string]map[string]struct {
			Offset float64 `json:"offset"`
			At     float64 `json:"at"`
		} `json:"shelves"`
	}
	if err := json.Unmarshal([]byte(text), &raw, jsonopt.Read()); err != nil {
		return doc
	}
	if raw.Version != DocumentVersion {
		return doc
	}

	for shelfKey, books := range raw.Shelves {
		trimmedKey := strings.TrimSpace(shelfKey)
		if trimmedKey == "" {
			continue
		}
		normalized := map[string]Entry{}
		for bookID, entry := range books {
			trimmedID := strings.TrimSpace(bookID)
			value, ok := normalizeEntry(entry.Offset, entry.At)
			if trimmedID != "" && ok {
				normalized[trimmedID] = value
			}
		}
		if len(normalized) > 0 {
			doc.Shelves[trimmedKey] = normalized
		}
	}

	return doc
}

// ParseStrict is Parse but rejects text that is not valid JSON. Write paths use
// it: lenient-parsing corrupt input into an empty document, then merging, would
// silently wipe the writer's stored namespace. An empty string is allowed and
// yields an empty document.
func ParseStrict(text string) (Document, error) {
	if strings.TrimSpace(text) != "" && !jsontext.Value(text).IsValid() {
		return New(), util.NewError("reading progress document is not valid JSON")
	}
	return Parse(text), nil
}

// Serialize renders the document as the compact JSON the frontend also produces.
func Serialize(doc Document) (string, error) {
	if doc.Version == 0 {
		doc.Version = DocumentVersion
	}
	if doc.Shelves == nil {
		doc.Shelves = map[string]map[string]Entry{}
	}
	bs, err := json.Marshal(doc, jsonopt.DiskCompact())
	if err != nil {
		return "", util.Errorf("%w", err)
	}
	return string(bs), nil
}

// Clone returns a deep copy so callers can mutate freely without touching the
// input — the projection and merge both build their result this way, which is
// what makes them safe to call repeatedly on the same document.
func (d Document) Clone() Document {
	out := Document{Version: d.Version, Shelves: make(map[string]map[string]Entry, len(d.Shelves))}
	if out.Version == 0 {
		out.Version = DocumentVersion
	}
	for shelfKey, books := range d.Shelves {
		out.Shelves[shelfKey] = cloneBooks(books)
	}
	return out
}

// ResolveShelf maps a stable book id to the real shelf id that owns it, or
// reports ok=false when no real shelf in the desktop library holds that book.
type ResolveShelf func(bookID string) (realShelfID string, ok bool)

// Project folds the reader's synthetic-shelf progress onto real shelves.
//
// For every book the reader recorded under readerShelfID, it asks resolve which
// real shelf holds that stable book id and, when one does, copies the entry to
// shelves[realShelfID][bookID] when the entry is newer (by At) than whatever is
// there. Entries under readerShelfID are always kept: an unresolved book (removed,
// or a loose book outside every shelf) stays there untouched, so a later import
// can pick it up on the next projection. The projection is therefore idempotent:
// running it again changes nothing once every resolvable book has been folded in.
func Project(doc Document, readerShelfID string, resolve ResolveShelf) Document {
	out := doc.Clone()
	readerShelfID = strings.TrimSpace(readerShelfID)
	readerBooks := out.Shelves[readerShelfID]
	if readerShelfID == "" || len(readerBooks) == 0 || resolve == nil {
		return out
	}

	for bookID, entry := range readerBooks {
		realShelfID, ok := resolve(bookID)
		if !ok {
			continue
		}
		realShelfID = strings.TrimSpace(realShelfID)
		if realShelfID == "" || realShelfID == readerShelfID {
			continue
		}
		dst := out.Shelves[realShelfID]
		if dst == nil {
			dst = map[string]Entry{}
			out.Shelves[realShelfID] = dst
		}
		if cur, exists := dst[bookID]; !exists || entry.At >= cur.At {
			dst[bookID] = entry
		}
	}

	return out
}

// MergeNewest overlays one process's write onto the on-disk document, keeping the
// newer entry (by At) for every book present in both. Books only on disk are
// preserved; books only in the write are added. This is the single reconciliation
// both the desktop and the reader use: because every entry is timestamped and a
// reset is a tombstone rather than a deletion, neither process can clobber the
// other's concurrent update, whichever shelf namespace it wrote — so both may
// write the same real shelf safely. On an equal timestamp the incoming write
// wins, which only ever happens when it carries the same value it just read.
func MergeNewest(disk, write Document) Document {
	out := disk.Clone()
	for shelfKey, books := range write.Shelves {
		shelfKey = strings.TrimSpace(shelfKey)
		if shelfKey == "" || len(books) == 0 {
			continue
		}
		dst := out.Shelves[shelfKey]
		if dst == nil {
			dst = map[string]Entry{}
			out.Shelves[shelfKey] = dst
		}
		for bookID, entry := range books {
			bookID = strings.TrimSpace(bookID)
			if bookID == "" {
				continue
			}
			if cur, exists := dst[bookID]; !exists || entry.At >= cur.At {
				dst[bookID] = entry
			}
		}
	}
	return out
}

// cloneBooks deliberately keeps the make+loop rather than maps.Clone. Document.Clone
// promises callers may mutate the returned maps freely, and a shelf decoded from a
// JSON "shelf": null yields a nil books map here — maps.Clone(nil) returns nil, so a
// caller that writes into the cloned shelf without a nil guard would panic. The
// make guarantees a non-nil map and preserves that contract unconditionally.
func cloneBooks(books map[string]Entry) map[string]Entry {
	out := make(map[string]Entry, len(books))
	for bookID, entry := range books {
		out[bookID] = entry
	}
	return out
}

// normalizeEntry keeps only the entries the frontend would keep. The offset must
// be a non-negative integer within the JSON-safe range; a non-integer or
// out-of-range timestamp is treated as unknown (0, the oldest) rather than
// dropping the entry. An entry with a zero offset survives only when it carries a
// timestamp — that is a reset tombstone; a bare {0,0} is nothing and is dropped.
func normalizeEntry(offset, at float64) (Entry, bool) {
	normalizedOffset, ok := normalizeNumber(offset)
	if !ok {
		return Entry{}, false
	}
	normalizedAt, ok := normalizeNumber(at)
	if !ok {
		normalizedAt = 0
	}
	if normalizedOffset == 0 && normalizedAt == 0 {
		return Entry{}, false
	}
	return Entry{Offset: normalizedOffset, At: normalizedAt}, true
}

// normalizeNumber accepts a non-negative, integral, JSON-safe value.
func normalizeNumber(value float64) (int64, bool) {
	if math.IsNaN(value) || math.IsInf(value, 0) {
		return 0, false
	}
	if value < 0 || value > maxSafeInteger {
		return 0, false
	}
	if value != math.Trunc(value) {
		return 0, false
	}
	return int64(value), true
}
